### PR TITLE
Allow a plan name to be displayed properly on the post-purchase authorization screen

### PR DESCRIPTION
**Pull request by jboland88**
## Proposed Changes

* This PR updates the post-connection authorization screen for Jetpack to correct a bug when a plan (not an individual product) was purchaed.

## Testing Instructions
* Start with a new Jetpack site
* From My Jetpack, while disconnected, click "Learn More" on the backup card and click to "Get Security"
* Without this patch, try completing the purchase - you should get an error screen just after checkout
* With this patch applied locally, copy and paste the checkout URL and try the checkout again in your local environment. It should complete the purchase and allow you to connect Jetpack afterwards

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/jetpack-connect/authorize.js [hunk 1]
```diff
@@ -888,7 +889,14 @@ export class JetpackAuthorize extends Component {
...
 		const product = getProductFromSlug( productSlug );
-		const productName = getJetpackProductDisplayName( product );
+		let productName = '';
+
+		if ( product?.product_name ) {
+			productName = getJetpackProductDisplayName( product );
+		} else {
+			productName = getPlan( productSlug )?.getTitle();
+		}
+
```
Comment (CodeyGuyDylan) [inline client/jetpack-connect/authorize.js]: I'd probably shorten this into a ternary 
```suggestion
        const productName = product?.product_name ? getJetpackProductDisplayName( product ) : getPlan( productSlug )?.getTitle();
```
but that is a personal preference

File: client/jetpack-connect/authorize.js [hunk 2]
```diff
@@ -888,7 +889,14 @@ export class JetpackAuthorize extends Component {
 		const productSlug = searchParams.get( 'productSlug' );
 		const siteSlug = searchParams.get( 'fromSiteSlug' );
 		const product = getProductFromSlug( productSlug );
-		const productName = getJetpackProductDisplayName( product );
+		let productName = '';
+
+		if ( product?.product_name ) {
+			productName = getJetpackProductDisplayName( product );
+		} else {
+			productName = getPlan( productSlug )?.getTitle();
```
Comment (CodeyGuyDylan) [inline client/jetpack-connect/authorize.js]: Could this be turned into a utility type function like `getJetpackProductDisplayName`? Might be useful elsewhere

**Review by CodeyGuyDylan — CHANGES_REQUESTED** [review]
Changes tested well, added a few suggestions, not necessarily blocking but I'll request changes just so we can discuss 😄

File: client/jetpack-connect/authorize.js (hunk 2 above)
Comment (jboland88) [reply client/jetpack-connect/authorize.js]: Nice idea - I'll look to add it in the same place as `getJetpackProductDisplayName `. Maybe we call it `getJetpackProductOrPlanDisplayName `?

File: packages/calypso-products/src/get-jetpack-product-or-plan-display-name.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import type { Product } from './types';
+import type { TranslateResult } from 'i18n-calypso';
+
+/**
+ * Get Jetpack product display name based on the product purchase object.
+ */
+export function getJetpackProductOrPlanDisplayName(
+	product: Product
+): TranslateResult | undefined {
+	return product?.product_name
```
Comment (robertsreberski) [inline packages/calypso-products/src/get-jetpack-product-or-plan-display-name.ts]:
```suggestion
	return product.product_name
```

product cannot be `undefined` or `null` at this point

File: client/jetpack-connect/authorize.js [hunk 4]
```diff
@@ -888,7 +888,7 @@ export class JetpackAuthorize extends Component {
 		const productSlug = searchParams.get( 'productSlug' );
 		const siteSlug = searchParams.get( 'fromSiteSlug' );
 		const product = getProductFromSlug( productSlug );
-		const productName = getJetpackProductDisplayName( product );
+		const productName = getJetpackProductOrPlanDisplayName( product );
```
Comment (robertsreberski) [inline client/jetpack-connect/authorize.js]: After introducing that function, `productName` can implicitly be `undefined`. That might lead to visual bugs in the text displayed to user.

We should either clean up the types (if we can be certain product name will be defined) or guard that edge case.

**Review by robertsreberski — COMMENTED** [review]
There are some typescript (hopefully only typing) nuances that we could improve. Besides that, lgtm

**Review by CodeyGuyDylan — APPROVED** [review]
LGTM! Tested well

**Pull request merged**